### PR TITLE
Record: XSA-all + GPTQ + FA3 dtype fix — val_bpb 1.1161 (3-seed mean)

### PR DESCRIPTION
**Pull request by G3sparky**
## Record Submission

**val_bpb: 1.1220** (sliding window, stride=64) | ~15.9 MB artifact | 8×H100 SXM, 600s

### Key Contribution

**FA3 dtype compatibility wrapper** — enables Flash Attention 3 Hopper kernels on PyTorch 2.5.1 which lacks auto-casting for FA3 calls. Simple 5-line wrapper that casts to bf16 when needed.

### Architecture

Built on the PR #1019 stack:
- 11L XSA on all layers
- Full Hessian GPTQ with AR self-gen calibration
- BigramHash 3072×112
- EMA (decay=0.997)
- LeakyReLU(0.5)²

### Results

| Steps | ms/step | Sliding BPB (s64) |
|-------|---------|-------------------|
| 6,244 | 96ms | **1.1220** |

8×H100 80GB SXM (Vast.ai), PyTorch 2.5.1+cu124, Flash Attention 3 compiled from source.

**Authors:** Gavin Saunders & Tron

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Record: XSA-all + GPTQ + FA3 dtype fix (val_bpb: 1.1220)

**BPB:** 1.1220 | **Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

**What I found in the code** (head SHA `7a94e720c629`, file `records/track_10min_16mb/2026-04-09_XSA_GPTQ_FA3_dtype_fix/train_gpt.py`):

Static code review found no TTT adaptation function, no SLOT optimization loop, no n-gram-cache class, and no pre-quant val-token fine-tune. The eval path uses the standard sliding-window stride-64 pattern. The submission is a pure-neural architecture iteration on the standard SP1024/SP4096/SP8192 baseline.

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 0.06s, dim=512, layers=11, vocab=1024, code=102083 B, SMOKE_TEST_PASS

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the classification pass — this looks like a clean pure-neural iteration on the standard baseline.

*Auto-classification caveat:* this review was drafted by the AST-based classifier. If there's a non-standard eval mechanism (logit postprocessing, hedge mixing, etc.) that I missed because it's factored into a helper file or a non-standard function name, please flag it and I'll re-run the audit manually.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 0.06s, dim=512, layers=11, vocab=1024, code=102083 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*

**Review by G3sparky — COMMENTED** [review]
Reviewed

File: records/track_10min_16mb/2026-04-09_XSA_GPTQ_FA3_dtype_fix/submission.json [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  "seed_results": {
+    "1337": {
+      "val_bpb": 1.1220,
+      "steps": 6244,
+      "step_avg_ms": 96.0
+    }
+  },
+  "hardware": "8xH100 80GB SXM (Vast.ai)",
+  "pytorch_version": "2.5.1+cu124",
+  "cuda_version": "13.0",
```
[the remark this answers is not in the document]
Comment (G3sparky) [reply records/track_10min_16mb/2026-04-09_XSA_GPTQ_FA3_dtype_fix/submission.json]: Good catch — the JSON should say cuda_version 12.4 to match cu124. Will fix in the next push.

File: records/track_10min_16mb/2026-04-09_XSA_GPTQ_FA3_dtype_fix/README.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+MUON_MOMENTUM=0.99 MUON_MOMENTUM_WARMUP_START=0.92 \
+MUON_MOMENTUM_WARMUP_STEPS=1500 WARMDOWN_ITERS=3000 \
+ITERATIONS=9000 MAX_WALLCLOCK_SECONDS=600 EVAL_STRIDE=64 \
+torchrun --standalone --nproc_per_node=8 train_gpt.py
+```
+
+## Hardware
+
+- 8× NVIDIA H100 80GB HBM3 SXM (Vast.ai, Nebraska)
+- PyTorch 2.5.1+cu124, CUDA 13.0
```
[the remark this answers is not in the document]
Comment (G3sparky) [reply records/track_10min_16mb/2026-04-09_XSA_GPTQ_FA3_dtype_fix/README.md]: Agreed, will correct to CUDA 12.4 to match the PyTorch build tag.

File: records/track_10min_16mb/2026-04-09_XSA_GPTQ_FA3_dtype_fix/train_gpt.py [hunk 3]
```diff
@@ -0,0 +1,2139 @@
...
+            param_name = name + ".weight"
+            cols = module.weight.shape[1]
+            hessians[param_name] = torch.zeros(cols, cols, dtype=torch.float32, device='cpu')
+            def make_hook(pname):
+                def hook_fn(module, input, output):
+                    x = input[0].detach().float()
+                    if x.ndim == 3:
+                        x = x.reshape(-1, x.shape[-1])
+                    hessians[pname] += (x.T @ x).cpu()
+                return hook_fn
```
[the remark this answers is not in the document]
Comment (G3sparky) [reply records/track_10min_16mb/2026-04-09_XSA_GPTQ_FA3_dtype_fix/train_gpt.py:1123]: Acknowledged. The Hessian accumulation on GPU with final .cpu() transfer is inherited from the upstream baseline. In practice the GPTQ calibration phase is < 12s on 8xH100 so the sync overhead is negligible, but keeping the Hessians on GPU until the end and doing a single bulk transfer would be cleaner. Will consider for the next iteration.

File: records/track_10min_16mb/2026-04-09_XSA_GPTQ_FA3_dtype_fix/train_gpt.py [hunk 4]
```diff
@@ -0,0 +1,2139 @@
...
+        code_bytes = len(code.encode("utf-8"))
+        log0(f"Serialized model int6+lzma: {quant_file_bytes} bytes")
+        log0(f"Total submission size int6+lzma: {quant_file_bytes + code_bytes} bytes")
+    if distributed:
+        dist.barrier()
+    with open("final_model.int6.ptz", "rb") as f:
+        quant_blob_disk = f.read()
+    quant_state = torch.load(
+        io.BytesIO(lzma.decompress(quant_blob_disk)),
+        map_location="cpu",
```
[the remark this answers is not in the document]
Comment (G3sparky) [reply records/track_10min_16mb/2026-04-09_XSA_GPTQ_FA3_dtype_fix/train_gpt.py:2063]: Fair point. Adding weights_only=True where applicable. The deserialization path only loads tensor data so weights_only=True is safe and removes the FutureWarning.

**Comment by G3sparky** [comment]
Hi @cocohearts — this PR is ready for review. All Copilot feedback addressed (CUDA version fixed, 3-seed results added, torch.load safety noted). 3-seed mean: **1.1161** (std 0.0009). All artifacts under 16MB. Thanks!

**Comment by G3sparky** [comment]
Superseded by #1858 (Neural-Only val_bpb 1.0810, 3-seed mean — ties leaderboard leader). Closing as the canonical entry has moved on. Thanks!
